src/eachrow.jl

```diff
@@ -181,14 +182,139 @@ julia> df2 = @eachrow df begin
            :colX = cols(varB) == 2 ? pi * cols(varA) : cols(varB)
        end
 3×3 DataFrame
-│ Row │ A     │ B     │ colX    │
-│     │ Int64 │ Int64 │ Float64 │
-├─────┼───────┼───────┼─────────┤
-│ 1   │ 1     │ 2     │ 3.14159 │
-│ 2   │ 2     │ 1     │ 1.0     │
-│ 3   │ 3     │ 2     │ 9.42478 │
+ Row │ A      B      colX
+     │ Int64  Int64  Float64
+─────┼───────────────────────
+   1 │     1      2  3.14159
+   2 │     2      1  1.0
+   3 │     3      2  9.42478
 ```
 """
 macro eachrow(d, body)
     esc(eachrow_helper(d, body, false))
 end
+
+function eachrow!_helper(df, body)
+    # @deprecate cannot be used because eachrow is a macro, and the @warn should not be in
+    # eachrow itself because then it will be displayed when the macro is evaluated.
+    e_body, e_newcols = eachrow_find_newcols(body, Any[])
+    _df = gensym()
+    quote
+        let $_df = $df
+            local _N = nrow($_df)
+            local _DF = @transform!($_df, $(e_newcols...))
+            $(with_helper(:_DF, :(for row = 1:_N
+                $(eachrow_replace(e_body))
+            end)))
+            _DF
+        end
+    end
+end
+
+"""
+    @eachrow!(d, body)
+
+Act on each row of a data frame, similar to
+
+```
+for row in eachrow(d)
+    ...
+end
+```
+
+Includes support for control flow and `begin end` blocks. Since the
+"environment" induced by `@eachrow df` is implicitly a single row of `d`,
+use regular operators and comparisons instead of their elementwise counterparts
+as in `@with`. Note that the scope within `@eachrow!` is a hard scope.
+
+`eachrow!` also supports special syntax for allocating new columns. The syntax
+`@newcol x::Vector{Int}` allocates a new column `:x` with an `Vector` container
+with eltype `Int`.This feature makes it easier to use `eachrow` for data
+transformations. `_N` is introduced to represent the length of the dataframe,
```

Comment:
is this what it was to mean
```suggestion
transformations. `_N` is introduced to represent the number of rows of the data frame,
```